### ISSUE TITLE
[clang][Sema] Invalid -Wshadow diagnostics in friend functions defined inside the class

### ISSUE THREAD

**Issue by brunodf-gf** [llvm/llvm-project#221190]
Clang raises a -Wshadow warning for the function parameter `x` in this C++ friend function:

```c++
class Foo {
    int x;
    friend bool operator==(const Foo& f, int x) {
        return f.x == x;
    }
};
```

Godbolt link: https://godbolt.org/z/dnr1sEaWM

Practically, I would argue there is no shadowing: the field `x` is not accessible inside the friend function (if you try to access it, you get "error: invalid use of non-static data member"). Gcc also does not give a warning for this case.

**Comment by AhmedKamel10** [comment]
Hi @brunodf-gf, I read the issue and did some research and I think I found a good starting point to work on it. Working on the PR rn.

**Comment by AhmedKamel10** [comment]
Created a PR fixing the issue.
#221249
